tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/parallel/RetryAnalyzer.java

```diff
@@ -0,0 +1,40 @@
+package org.alfresco.elasticsearch.parallel;
+
+import java.util.ConcurrentModificationException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.IRetryAnalyzer;
+import org.testng.ITestResult;
+
+/**
+ * Sometimes REST tests fail when run in parallel with:
+ * <pre>
+ *     java.lang.IllegalStateException: Invalid use of BasicClientConnManager: connection still allocated.
+ *     Make sure to release the connection before allocating another one.
+ * </pre>
+ * Other times restassured fails with <code>java.util.ConcurrentModificationException</code>.
+ * If we detect a test failed due to these timing issues then we rerun it up to three times.
```

Comment:
Those exceptions are thrown due to shared user and site?

Reply:
These exceptions seem to be due to multiple tests using RestAssured in parallel. I ran the tests about 30 times locally (with various settings) and these are the only two exceptions I saw which were not application issues.

Reply:
RestAssured probably is not entirely thread-safe, so I think you're right. Thanks for explaining.